pkg/ingester/ingester.go

```diff
@@ -2299,6 +2301,122 @@ func (i *Ingester) RingHandler() http.Handler {
+	if err := i.checkRunning(); err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+
+	level.Info(i.logger).Log("msg", "processing request to add backfill file")
+
+	f, err := os.CreateTemp("", "")
+	if err != nil {
+		return errors.Wrap(err, "creating temp file")
+	}
+	defer func() {
+		_ = f.Close()
+		if err := os.Remove(f.Name()); err != nil {
+			level.Warn(i.logger).Log("msg", "failed to remove temporary file", "file", f.Name(), "err", err)
+		}
+	}()
+
+	var tenantID int
+	var blockID string
+	var pth string
+	var chunk int
+	var bytesWritten int64
+	for {
+		req, err := stream.Recv()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return errors.Wrap(err, "failed to receive from gRPC stream")
+		}
+
+		if blockID == "" {
+			tenantID = int(req.TenantId)
+			blockID = req.BlockId
+			pth = req.Path
+			level.Info(i.logger).Log("msg", "adding file to backfill", "tenantId",
+				tenantID, "blockId", blockID, "path", pth, "chunkLength",
+				len(req.Chunk))
+			chunk = 1
+		} else {
+			chunk++
+		}
+
+		if _, err := f.Write(req.Chunk); err != nil {
+			return errors.Wrap(err, "failed to write to file")
+		}
+		bytesWritten += int64(len(req.Chunk))
+	}
+	if chunk == 0 {
+		level.Warn(i.logger).Log("msg", "no backfill file content was sent")
+		return stream.SendAndClose(&mimirpb.AddBackfillFileResponse{})
+	}
+
+	level.Info(i.logger).Log("msg", "finished writing chunks to backfill file", "tenantId",
+		tenantID, "blockId", blockID, "path", pth, "chunks", chunk, "bytesWritten", bytesWritten)
+
+	if _, err := f.Seek(0, 0); err != nil {
+		return errors.Wrap(err, "failed seeking in file")
+	}
+
+	dst := path.Join("uploads", blockID, pth)
+	level.Info(i.logger).Log("msg", "uploading backfill file to bucket", "tenantId", tenantID,
+		"destination", dst)
+	bkt := bucket.NewUserBucketClient(string(tenantID), i.bucket, i.limits)
+	defer bkt.Close()
+	if err := bkt.Upload(ctx, dst, f); err != nil {
+		return errors.Wrap(err, "failed uploading backfill file to bucket")
+	}
+
+	return stream.SendAndClose(&mimirpb.AddBackfillFileResponse{})
+}
+
+func (i *Ingester) FinishBackfill(ctx context.Context, req *mimirpb.FinishBackfillRequest) (*mimirpb.FinishBackfillResponse, error) {
+	if err := i.checkRunning(); err != nil {
+		return nil, err
+	}
+
+	level.Info(i.logger).Log("msg", "processing request to finish backfill", "tenantId",
+		req.TenantId, "blockId", req.BlockId, "files", len(req.Files))
+
+	tenantID := strconv.Itoa(int(req.TenantId))
+	bkt := bucket.NewUserBucketClient(tenantID, i.bucket, i.limits)
+	defer bkt.Close()
+
+	for _, fpath := range req.Files {
+		if path.Base(fpath) == "meta.json" {
+			continue
+		}
+
+		src := path.Join(tenantID, "uploads", req.BlockId, fpath)
+		dst := path.Join(tenantID, req.BlockId, fpath)
+		level.Info(i.logger).Log("msg", "moving file in bucket", "src", src, "dst", dst)
+		if err := bkt.Move(ctx, src, dst); err != nil {
```

Comment:
I'm just reading this randomly, so forgive me if I'm missing/misunderstanding something fundamental: why is a temporary upload needed that is then copied?

From my attempt at understanding backfilling it seems like the primary problem is a new ingestion path derived from already complete but untrusted blocks rather than series which are packed into blocks server-side. Is the intent here to find a place to gather the complete state of the block, open it up to an arbitrary transform/validation, and then move it?

If in-progress uploads are protected by the lack of a `meta.json` already, could there possibly be a way to perform the validation/transformation file by file while sending it directly to its intended final location instead to avoid the intermediary step?

Reply:
@pracucci @pstibrany WDYT about @andyasp's question regarding whether we can copy directly to the final destination, and just defer writing meta.json until we're ready?

Reply:
Instead of deferring meta.json file, I would think about using some marker file for in-progress uploads. This would be more explicit for other components (ie. compactor can blocks that are being uploaded, or it can detect that upload hasn't progressed in a while, if we keep in-progress marker file updated with timestamp)

Client may do upload of files in any order, and we cannot rely on correct client behaviour. Delaying upload in other component (eg. keeping file in memory or local disk) risks that we lose uploaded meta.json file before it is stored to the storage.

Reply:
In other words, I agree that we can upload directly to target block destination, but we also need a way to detect which blocks are still being uploaded.

Reply:
@pstibrany after the client (typically mimirtool) finishes sending files, it will make a request to finish the block backfill. As part of handling this request, Mimir could write the block meta.json file. If it does fail, the client will receive the error back of course. Wouldn't this scheme work?

Reply:
Are you saying that client would only send `meta.json` in the "finish the block" method?

Reply:
@pstibrany that's what I have in mind, yeah.

Reply:
I've implemented what I discussed with @pstibrany above - i.e., upload block files directly to the destination, but only upload meta.json as the very last step (during session completion).